api/indexer/indexer.go

```diff
@@ -8,230 +8,252 @@ import (
 
 	"github.com/ava-labs/hypersdk/chain"
 	"github.com/ava-labs/hypersdk/event"
-	"github.com/ava-labs/hypersdk/fees"
-	"github.com/ava-labs/hypersdk/internal/cache"
 	"github.com/ava-labs/hypersdk/internal/pebble"
 )
 
 const maxBlockWindow uint64 = 1_000_000
 
+var (
+	blockEntryKeyPrefix  = []byte("block")
+	latestBlockHeightKey = []byte("latestblockheight")
+)
+
 var errBlockNotFound = errors.New("block not found")
 
 var _ event.Subscription[*chain.ExecutedBlock] = (*Indexer)(nil)
 
 type Indexer struct {
 	blockDB            *pebble.Database // height -> block bytes
-	blockIDToHeight    *cache.FIFO[ids.ID, uint64]
-	blockHeightToBlock *cache.FIFO[uint64, *chain.ExecutedBlock]
-	blockWindow        uint64 // Maximum window of blocks to retain
-	lastHeight         atomic.Uint64
+	cachedBlocks       buffer.Queue[uint64]
+	blockIDToHeight    map[ids.ID]uint64
+	blockHeightToBlock map[uint64]*chain.ExecutedBlock
+	txCache            map[ids.ID]cachedTransaction
+	evictedBlocks      []uint64 // list of blocks that are queued for deletion.
+	blockWindow        uint64   // Maximum window of blocks to retain
+	lastHeight         uint64
 	parser             chain.Parser
 
-	// ID -> timestamp, success, units, fee, outputs
-	txDB *pebble.Database
+	// synchronization mutex between r/w
+	mu sync.RWMutex
+}
+
+type cachedTransaction struct {
+	// blk is the block that contains the transaction
+	blk *chain.ExecutedBlock
+	// index is the position where the transaction appear within the block.
+	index int
 }
 
 func NewIndexer(path string, parser chain.Parser, blockWindow uint64) (*Indexer, error) {
 	if blockWindow > maxBlockWindow {
 		return nil, fmt.Errorf("block window %d exceeds maximum %d", blockWindow, maxBlockWindow)
 	}
-	txDB, err := pebble.New(filepath.Join(path, "tx"), pebble.NewDefaultConfig(), prometheus.NewRegistry())
-	if err != nil {
-		return nil, err
-	}
 	blockDB, err := pebble.New(filepath.Join(path, "block"), pebble.NewDefaultConfig(), prometheus.NewRegistry())
 	if err != nil {
 		return nil, err
 	}
 
-	blockIDCache, err := cache.NewFIFO[ids.ID, uint64](int(blockWindow))
-	if err != nil {
-		return nil, err
-	}
-	blockCache, err := cache.NewFIFO[uint64, *chain.ExecutedBlock](int(blockWindow))
-	if err != nil {
-		return nil, err
-	}
 	i := &Indexer{
 		blockDB:            blockDB,
-		blockIDToHeight:    blockIDCache,
-		blockHeightToBlock: blockCache,
+		blockIDToHeight:    make(map[ids.ID]uint64, int(blockWindow)),
+		blockHeightToBlock: make(map[uint64]*chain.ExecutedBlock, int(blockWindow)),
+		txCache:            make(map[ids.ID]cachedTransaction),
 		blockWindow:        blockWindow,
 		parser:             parser,
-		txDB:               txDB,
+		lastHeight:         math.MaxUint64,
+	}
+
+	i.cachedBlocks, err = buffer.NewBoundedQueue(int(blockWindow), i.evictBlock)
+	if err != nil {
+		return nil, err
 	}
+
 	return i, i.initBlocks()
 }
 
+func (i *Indexer) evictBlock(blockHeight uint64) {
```

Comment:
do we rely on the assumption that the lock is held whenever we insert into the bounded buffer, so that the lock is held when this is called?

The indirection here makes this a bit hard to follow, so I think we should either try and simplify this or clearly document what the invariants are here to ensure there's no race conditions.

Reply:
I don't think it's any different than FIFO, and I added a description so to reduce confusion.